File: csharp.xUnit/GildedRose/GildedRose.cs
using System.Collections.Generic;

namespace GildedRoseKata
{
    /// <summary>
    /// Represents the Gilded Rose inventory system responsible for updating item quality and sell-in values.
    /// </summary>
    public class GildedRose
    {
        private readonly IList<Item> _items;

        public GildedRose(IList<Item> items)
        {
            _items = items;
        }

        /// <summary>
        /// Updates the quality and sell-in values of all items in the inventory according to business rules.
        /// </summary>
        public void UpdateQuality()
        {
            foreach (var item in _items)
            {
                UpdateItemQuality(item);
            }
        }

        /// <summary>
        /// Updates the quality and sell-in values for a single item.
        /// </summary>
        /// <param name="item">The item to update.</param>
        private static void UpdateItemQuality(Item item)
        {
            if (IsLegendaryItem(item))
            {
                return;
            }

            AdjustQualityBeforeSellIn(item);
            item.SellIn--;

            if (item.SellIn < 0)
            {
                AdjustQualityAfterSellIn(item);
            }
        }

        private static void AdjustQualityBeforeSellIn(Item item)
        {
            if (IsAgedBrie(item))
            {
                IncreaseQuality(item);
            }
            else if (IsBackstagePass(item))
            {
                IncreaseQuality(item);
                if (item.SellIn < 11) IncreaseQuality(item);
                if (item.SellIn < 6) IncreaseQuality(item);
            }
            else
            {
                DecreaseQuality(item);
            }
        }

        private static void AdjustQualityAfterSellIn(Item item)
        {
            if (IsAgedBrie(item))
            {
                IncreaseQuality(item);
            }
            else if (IsBackstagePass(item))
            {
                item.Quality = 0;
            }
            else
            {
                DecreaseQuality(item);
            }
        }

        private static void IncreaseQuality(Item item)
        {
            if (item.Quality < 50)
            {
                item.Quality++;
            }
        }

        private static void DecreaseQuality(Item item)
        {
            if (item.Quality > 0)
            {
                item.Quality--;
            }
        }

        private static bool IsAgedBrie(Item item) => item.Name == "Aged Brie";
        private static bool IsBackstagePass(Item item) => item.Name == "Backstage passes to a TAFKAL80ETC concert";
        private static bool IsLegendaryItem(Item item) => item.Name == "Sulfuras, Hand of Ragnaros";
    }
}

File: Java-Spock/src/main/java/com/gildedrose/GildedRose.java
package com.gildedrose;

class GildedRose {
    private static final String AGED_BRIE = "Aged Brie";
    private static final String BACKSTAGE_PASSES = "Backstage passes to a TAFKAL80ETC concert";
    private static final String SULFURAS = "Sulfuras, Hand of Ragnaros";
    private static final int MAX_QUALITY = 50;

    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {
        for (Item item : items) {
            updateItem(item);
        }
    }

    private void updateItem(Item item) {
        if (isSulfuras(item)) {
            return;
        }

        updateQualityBeforeSellIn(item);
        item.sellIn -= 1;

        if (item.sellIn < 0) {
            updateQualityAfterSellIn(item);
        }
    }

    private void updateQualityBeforeSellIn(Item item) {
        if (isAgedBrie(item)) {
            increaseQuality(item);
            return;
        }

        if (isBackstagePass(item)) {
            increaseQuality(item);
            if (item.sellIn < 11) {
                increaseQuality(item);
            }
            if (item.sellIn < 6) {
                increaseQuality(item);
            }
            return;
        }

        decreaseQuality(item);
    }

    private void updateQualityAfterSellIn(Item item) {
        if (isAgedBrie(item)) {
            increaseQuality(item);
            return;
        }

        if (isBackstagePass(item)) {
            item.quality = 0;
            return;
        }

        decreaseQuality(item);
    }

    private void increaseQuality(Item item) {
        if (item.quality < MAX_QUALITY) {
            item.quality += 1;
        }
    }

    private void decreaseQuality(Item item) {
        if (item.quality > 0) {
            item.quality -= 1;
        }
    }

    private boolean isAgedBrie(Item item) {
        return AGED_BRIE.equals(item.name);
    }

    private boolean isBackstagePass(Item item) {
        return BACKSTAGE_PASSES.equals(item.name);
    }

    private boolean isSulfuras(Item item) {
        return SULFURAS.equals(item.name);
    }
}
